core/classes/Misc/MentionsParser.php

```diff
@@ -11,49 +11,78 @@
  */
 class MentionsParser
 {
+    private const USER_MENTIONS_REGEX = '/(?<!\/)@([A-Za-z0-9\-_!.]+)/';
+
+    /**
+     * Parse the given content to replace @username tags with [user]<id>[/user] bbcode.
+     *
+     * @param int    $author_id User ID of post/custom page creator.
+     * @param string $content   Post/custom page content.
+     *
+     * @return string Parsed post content.
+     */
+    public static function parse(int $author_id, string $content): string
+    {
+        $receipients = self::getRecipients($content, $author_id);
+
+        return self::replaceWithBbcode($content, $receipients);
+    }
+
+    public static function parseAndNotify(int $author_id, string $content, string $url, string $notificationType, LanguageKey $notificationTitle): string
+    {
+        $receipients = self::getRecipients($content, $author_id);
+
+        $notificationRecipients = array_filter($receipients, fn ($receipient) => $receipient->id !== $author_id);
+        $notificationRecipients = array_column($notificationRecipients, 'id');
+
+        $notification = new Notification(
+            $notificationType,
+            $notificationTitle,
+            // TODO: emails content - right now it will be plaintext and not use a template
+            $content,
+            $notificationRecipients,
+            $author_id,
+            null,
+            false,
+            $url,
+        );
+
+        $notification->send();
+
+        return self::replaceWithBbcode($content, $receipients);
+    }
+
+    /**
+     * Get users from the database based on the provided nicknames. Filters out users have blocked the author.
+     */
+    private static function getRecipients(string $content, int $author_id): array
+    {
+        preg_match_all(self::USER_MENTIONS_REGEX, $content, $matches);
+        $nicknames = $matches[1];
+
+        return DB::getInstance()->query(
```

Comment:
User should be able to tag the user despite being blocked thought but not alert

Reply:
fixed